[NFC][SYCL][L0] Add infrastructure to know backend of a platform

smaslov-intel opened the pull request:
In subsequent PR the "backend" knowledge will move from "plugin" to "platform"

jandres742 reviewed (commented):
+1

jchlanda reviewed (approved):
CUDA/HIP LGTM

jchlanda commented on sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp:
```diff
@@ -465,6 +465,11 @@ pi_result piPlatformGetInfo(pi_platform Platform, pi_platform_info ParamName,
 
+  case PIEXT_PLATFORM_INFO_BACKEND:
+    return getInfo<pi_platform_backend>(
+        param_value_size, param_value, param_value_size_ret,
```
```suggestion
        ParamValueSize, ParamValue, ParamValueSizeRet,
```

steffenlarsen commented on sycl/cmake/modules/AddSYCL.cmake:
```diff
@@ -50,7 +50,8 @@ function(add_sycl_plugin PLUGIN_NAME)
       ${ARG_LIBRARIES}
-      OpenCL-Headers
+      OpenCL-Headers # TODO: remove dependency on OpenCL
+      UnifiedRuntime-Headers
```
`UnifiedRuntime-Headers` seems to be added by sycl/plugins/unified_runtime/CMakeLists.txt, right? If I am reading sycl/plugins/CMakeLists.txt correctly, the `unified_runtime` plugin sub-directory is only added if the `level_zero` plugin in enabled, i.e.
```cmake
# level_zero plugin depends today on unified_runtime plugin
# and unified_runtime plugin is not an independent plugin, adding it explicitly
if ("level_zero" IN_LIST SYCL_ENABLE_PLUGINS)
  add_subdirectory(unified_runtime)
endif()
```
Since a lot more projects would depend on the UR headers after this, should we consider either enabling the `unified_runtime` plugin in more cases or move the logic for getting the headers?

smaslov-intel replied:
Good point! Yes, I made UR be added always now that all plugins depend on it (through PI->UR dependency).
When entire UR is moved into its separate repo we should consider greater restructure, like moving UR one level up.

smaslov-intel commented:
@intel/llvm-reviewers-runtime : hmm, is it true/correct that pi.h is included into every SYCL program?

> In file included from /__w/llvm/llvm/src/clang/test/CodeGenSYCL/sycl-add-opt-level-attrib.cpp:9:
**In file included from /__w/llvm/llvm/build/bin/../include/sycl/sycl.hpp:11:**
In file included from /__w/llvm/llvm/build/bin/../include/sycl/accessor.hpp:12:
In file included from /__w/llvm/llvm/build/bin/../include/sycl/atomic.hpp:12:
In file included from /__w/llvm/llvm/build/bin/../include/sycl/access/access.hpp:11:
In file included from /__w/llvm/llvm/build/bin/../include/sycl/detail/common.hpp:14:
In file included from /__w/llvm/llvm/build/bin/../include/sycl/detail/pi.hpp:19:
/__w/llvm/llvm/build/bin/../include/sycl/detail/pi.h:127:10: fatal error: 'ur_api.h' file not found
#include <ur_api.h>
         ^~~~~~~~~~
1 error generated.

smaslov-intel commented:
> @intel/llvm-reviewers-runtime : hmm, is it true/correct that pi.h is included into every SYCL program? Then we need UR on the same level as SYCL includes.
> 
> > In file included from /__w/llvm/llvm/src/clang/test/CodeGenSYCL/sycl-add-opt-level-attrib.cpp:9:
> > **In file included from /__w/llvm/llvm/build/bin/../include/sycl/sycl.hpp:11:**
> > In file included from /__w/llvm/llvm/build/bin/../include/sycl/accessor.hpp:12:
> > In file included from /__w/llvm/llvm/build/bin/../include/sycl/atomic.hpp:12:
> > In file included from /__w/llvm/llvm/build/bin/../include/sycl/access/access.hpp:11:
> > In file included from /__w/llvm/llvm/build/bin/../include/sycl/detail/common.hpp:14:
> > In file included from /__w/llvm/llvm/build/bin/../include/sycl/detail/pi.hpp:19:
> > /__w/llvm/llvm/build/bin/../include/sycl/detail/pi.h:127:10: fatal error: 'ur_api.h' file not found
> > #include <ur_api.h>
> > ^~~~~~~~~~
> > 1 error generated.

I have changed the PR to not use UR directly from PI, and added this translation
https://github.com/intel/llvm/pull/9067/files#diff-f1103e5442bf4f71ac0fb3cd9aa969042ee3e864e7b3e8c6f3255b7162ffadfeR168

In a subsequent PR we need to move UR to the top level.

smaslov-intel commented:
@steffenlarsen , @againull : kind ping

steffenlarsen reviewed (approved):
Looks good!
